proposals/kubeflow-versioning.md

```diff
@@ -0,0 +1,54 @@
+# Kubeflow Versioning
+
+## Motivation
+
+Regularize the Kubeflow release schedule.
+
+## Proposal
+
+In brief, this proposal would switch the release cycle to this:
+
+![Kubeflow Versioning](diagrams/kubeflow-versioning.png)
+
+The changes this would introduce are:
+
+- A new version of Kubeflow would be released every 2 months
+- The deadline would be strict; if something is not ready for a particular release, it will be bumped to the next one
+- Each version would be supported (i.e. bugs/security issues fixed) for 4 months, except for yearly Extended Support (ES) versions that would be supported for 16 months
+- Kubeflow would be versioned according to a YY.MM.N scheme, where YY and MM are the two-digit year and month of the release date, and N is the patch number. N starts at 0 for each YY.MM release, and is incremented by one for each patch release.
+
+This proposal would not change:
+
+- The release cycle of individual Kubeflow components
+
+## Motivation
+
+- Strict release schedule
+  - It is not easy to plan around the current release schedule, as it does not have hard deadlines.
+- Bumping components to subsequent releases
+  - Kubeflow components must not block releases for this proposal to work, otherwise Kubeflow releases can't be strict.
+- 2 month release cycle
+  - This proposal chooses a 2 month release cycle due to the fact that if a component of Kubeflow is bumped to a subsequent release, there should not be too long of a wait until it can be included in another release. Otherwise, there will be pressure to delay a release to include a new component release, which reverts back to the current schedule.
+- 4 month support cycle
+  - This length of time is enough that there are always 3 supported versions of Kubeflow, assuming a 2 month release cycle. This number of supported versions matches Kubernetes, where it has seemed to work well.
+- 16 month Extended Support cycle
+  - This length of time is long enough that distributors and consumers of Kubeflow that are not interested in a rapid release cycle or immediate new features are not forced to continuously upgrade Kubeflow versions.
+- YY.MM versioning scheme
+  - This scheme is proposed instead of semantic versioning due to the fact that each individual Kubeflow component may or may not have breaking changes in its release, and that information cannot be captured accurately in a single version number. It also helps enforce a strict release schedule, as releasing say 21.04 in May or June of 2021 is not very palatable.
```

Comment:
I am interested to know whether there is any alternatives: SEMVER is widely adopted by many products: https://semver.org/. The benefit of SEMVER is that MAJOR version bump is breaking change, while other version upgrade can be straightforward. It is preferable for big apps with many dependencies, it helps them to determine when and how to upgrade dependencies.

Having this YY.MM.N schema is basically replicating the information of release timestamp: this release time information can be found on kubeflow release note. And 4 months of support for any version is relatively short for company to act.

Taking the `Component Versioning Example` as example: Kubeflow release essentially is a collection of versioned components in a package. Alternative to YY.MM.N: We can upgrade the kubeflow SEMVER based on the Biggest change of components: If there is at least a component with MAJOR version bump, then we bump Kubeflow MAJOR version. Otherwise check whether there is MINOR version bump in components, if yes, then bump Kubeflow MINOR version.

## Exmaple

| Kubeflow Release |  A  |  B  |  C  |
| :--------------- | :-: | :-: | :-: |
| 1.1.0            | 1.0.0 | 0.5.0 | 1.3.0 |
| 1.2.0            | 1.1.0 | 0.6.0 | 1.4.0 |
| 1.2.1            | 1.1.0 | 0.6.1 | 1.4.1 |
| 2.0.0           | 2.0.0 | 0.6.1 | 1.5.0 |
| 2.1.0            | 2.0.0 | 0.7.0 | 1.5.1 |

Reply:
Thanks for the response. To be clear, the meat of this proposal is in the regularized release cadence; it would work just as well with semantic versioning.

That being said, I think the difficulty with semantic versioning here is that Kubeflow consists of enough components that there's likely to always be a breaking change in one of them. This would render semantic versioning semi-useless, since the answer to "is it a breaking change" would always be "yes". I'm also not sure how optional components would be handled. If an optional Kubeflow component included a breaking change, would that bump the major Kubeflow version?

> And 4 months of support for any version is relatively short for company to act.

I definitely agree, and that's why the proposal includes Extended Support releases that would allow companies to upgrade Kubeflow on a more relaxed, once a year schedule.

Reply:
Sounds good on providing a more relaxed, once a year schedule for Extended Support releases. 👍 

About breaking changes and SEMVER: I think your argument of breaking changes in each Kubeflow release makes sense. However, I hope that not every Kubeflow release contains major component upgrade: That means we need to maintain a long list of upgrade guide across the cloud providers. For example: We need to maintain a document about `How do you upgrade from v1 to v2, v2 to v3, v3 to v4, etc.`, we don't hope that Kubeflow components bring in breaking changes as frequently as monthly. So my suggestion is to schedule a `breaking change Kubeflow release` in annual or semi-annual frequency, where Kubeflow components can sign up for their breaking changes release.

My opinion towards optional components: the breaking change of optional component will probably be taken into account too, as long as they are listed under kubeflow/manifest.